torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -156,10 +153,55 @@ void PropagateShapeOnNodeByRunningIt(Node* node, const std::vector<TensorType*>&
 
   JIT_ASSERT(stack.size() == node->outputs().size());
   for (size_t i = 0; i < stack.size(); ++i) {
-    node->outputs()[i]->inferTypeFrom(stack[i].toTensor());
+    // some ops may have mixed tensor/primitive outputs
+    // for primitives, we don't need to change the type because it is already
+    // its most constrained form.
+    if(stack[i].isTensor())
+      node->outputs()[i]->inferTypeFrom(stack[i].toTensor());
   }
 }
 
+// is it ok to try to run the op
+// If an input is a constant, then we assume that the input is valid
+// and we can try to run it.
+// Otherwise:
+// Integral typed _inputs_ are often an indicator that we're indexing into
+// a tensor, so we should special-case these ops in the shape propagation.
+// Additionally, passing in a zero representative tensor into an integer
+// division op causes divide-by-zero errors
+// _Outputs_ must be tensors or primtives
```

Comment:
Is there anything that's not a tensor or a primitive?

Reply:
Lists of stuff, tuples of stuff, (soon strings)